src/libslic3r/PrintObject.cpp

```diff
@@ -3271,32 +3271,38 @@ void PrintObject::discover_horizontal_shells()
                         }
                     }
 
-                    if (region_config.sparse_infill_density.value == 0 || region_config.ensure_vertical_shell_thickness.value == evstCriticalOnly) {
+                    float factor = 0.0f;
+                    if (region_config.sparse_infill_density.value == 0)
+                        factor = 1.0f;
+                    else if (region_config.ensure_vertical_shell_thickness.value == evstNone)
+                        factor = 0.5f;
+                    else if (region_config.ensure_vertical_shell_thickness.value == evstCriticalOnly)
+                        factor = 0.2f;
+                    if (factor > 0.0f) {
                         // if we're printing a hollow object we discard any solid shell thinner
                         // than a perimeter width, since it's probably just crossing a sloping wall
                         // and it's not wanted in a hollow print even if it would make sense when
                         // obeying the solid shell count option strictly (DWIM!)
-                        
+
                         // Orca: Also use the same strategy if the user has selected to reduce
                         // the amount of solid infill on walls. However reduce the margin to 20% overhang
                         // as we want to generate infill on sloped vertical surfaces but still keep a small amount of
                         // filtering. This is an arbitrary value to make this option safe
                         // by ensuring that top surfaces, especially slanted ones dont go **completely** unsupported
                         // especially when using single perimeter top layers.
-                        float margin = region_config.ensure_vertical_shell_thickness.value == evstCriticalOnly? float(neighbor_layerm->flow(frExternalPerimeter).scaled_width()) * 0.2f : float(neighbor_layerm->flow(frExternalPerimeter).scaled_width());
-                        Polygons too_narrow = diff(
-                            new_internal_solid,
-                            opening(new_internal_solid, margin, margin + ClipperSafetyOffset, jtMiter, 5));
+                        float    margin     = float(neighbor_layerm->flow(frExternalPerimeter).scaled_width()) * factor;
```

Comment:
@igiannakas 
I've identified a potential issue in the code. Previously, when ensure_vertical_shell_thickness was set to Critical, we would default to a 0.2 factor even if the user had set sparse_infill_density to 0. According to the original logic in Slic3r, we should use a factor of 1 to prevent creating extra solid infill when printing a hollow model. This approach seems logical to me, so I've adjusted the logic here to give sparse_infill_density precedence over the ensure_vertical_shell_thickness setting.

Please let me know if this change was made intentionally on your part.

Reply:
Yes that was intentional as it was causing top layers to be unsupported…

Reply:
<img width="1779" alt="image" src="https://github.com/SoftFever/OrcaSlicer/assets/59056762/97363251-f62c-4bd8-b17f-22a9e62a1999">

https://github.com/SoftFever/OrcaSlicer/pull/3235

towards the end of the thread

Reply:
@igiannakas Sorry I can't find the `sparse_infill_density` part.

Reply:
I'm referring a situation when user has these settings: sparse_infill_density = 0 and ensure_vertical_shell_thickness=Critical(previously further reduce solid infill option),